### PR TITLE
[Clang] Prevent assignment to captured structured bindings inside immutable lambda

### LINKED ISSUE #95081: Permitted assignment to read-only capture in not-mutable lambda
This program
```
struct B {
    int i;
};

int foo() {
    auto [x] = B{1};
    [x]() {
        x = 2;
    }();
    return x;
}
```
is invalid because not-mutable lambda modifies its read-only capture `x`, and it is properly rejected by GCC and MSVC, but Clang erroneously admits it. Online demo: https://gcc.godbolt.org/z/46o6jad8n

### PR DESCRIPTION
**Pull request by TilakChad**
For structured bindings, a call to getCapturedDeclRefType(...) was missing. This PR fixes that behavior and adds the related diagnostics too. 

This fixes https://github.com/llvm/llvm-project/issues/95081.

### CODE REVIEW

File: clang/lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -13299,7 +13293,18 @@ static NonConstCaptureKind isReferenceToNonConstCapture(Sema &S, Expr *E) {
 
   // The declaration must be a variable which is not declared 'const'.
   VarDecl *var = dyn_cast<VarDecl>(DRE->getDecl());
-  if (!var) return NCCK_None;
+  if (!var) {
+    // Bindings also can be captured by lambda in C++
+    BindingDecl *binding = dyn_cast<BindingDecl>(DRE->getDecl());
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaExpr.cpp:13298]:
```suggestion
    BindingDecl *Binding = dyn_cast<BindingDecl>(DRE->getDecl());
```

File: clang/lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -13299,7 +13293,18 @@ static NonConstCaptureKind isReferenceToNonConstCapture(Sema &S, Expr *E) {
...
+    if (!binding || binding->getType().isConstQualified())
+      return NCCK_None;
+
+    assert(S.getLangOpts().CPlusPlus && "BindingDecl outside of C++?");
+    assert(!isa<BlockDecl>(binding->getDeclContext()));
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaExpr.cpp:13303]: I would prefer you move the whole code _before_ the cast to Var (L13293)
Alternatively, first cast to ValueDecl, deal with the common const-qualified case, then cast to Var

**Review by cor3ntin — COMMENTED** [review]
I have a few nits but generally looks good, thanks for the fix

File: clang/lib/Sema/SemaExpr.cpp (hunk 1 above)
Comment (TilakChad) [reply clang/lib/Sema/SemaExpr.cpp]: Noted.

File: clang/lib/Sema/SemaExpr.cpp (hunk 2 above)
Comment (TilakChad) [reply clang/lib/Sema/SemaExpr.cpp]: I've added the commit based on your suggestion. 
Thank you for the review. :smile:

**Review by cor3ntin — APPROVED** [review]
LGTM, thanks

**Comment by TilakChad** [comment]
Thanks @cor3ntin for the review and the approval. 

I don't have commit access  as it's my very first PR. So, could you please merge this PR for me?

**Pull request merged**